_blogposts/2023-09-15-uncurried-mode.mdx

```diff
@@ -0,0 +1,127 @@
+---
+author: rescript-team
+date: "2023-09-15"
+title: Uncurried Mode
+badge: roadmap
+description: |
+  A tour of new capabilities coming to ReScript v11
+---
+
+> This is the fourth post covering new capabilities that'll ship in ReScript v11. You can check out the first post on [Better Interop with Customizable Variants](/blog/improving-interop), the second post on [Enhanced Ergonomics for Record Types](/blog/enhanced-ergonomics-for-record-types) and the third post on [First-class Dynamic Import Support](/blog/first-class-dynamic-import-support).
+
+## Introduction
+
+ReScript is a language that strives to keep its users free from experiencing runtime errors. Usually, when a program compiles, it will already do what the user described.
+But there is still a concept in the language that makes it easy to let some errors slip through. Currying! 
+
+Because of currying, partial application of functions is possible. That feature is always advertised as something really powerful. For instance, 
+
+```rescript
+let add = (a, b) => a + b
+let addFive = add(5)
+```
+
+is shorter than having to write all remaining parameters again 
+
+```rescript
+let add = (a, b) => a + b
+let addFive = (b) => add(5, b)
+```
+
+Also, currying is what enabled ReScript to have optional (`~param=?`) or default (`~param=true`) parameters. 
+
+This comes at a price though. Here are some examples to show the drawbacks of currying: 
+
+* Errors because of changed function signatures have their impact at the use site. Consider this example, where the signature of the onChange function is extended with
+a labeled argument 
+  ```diff
+    @react.component
+  -  let make = (~onChange: string => option<unit => unit>) => {
+  +  let make = (~onChange: (~a: int, string) => option<unit => unit>) => {
+      React.useEffect(() => {
+        // As partial application is allowed, there is no error here.
+        let cleanup = onChange("change") 
+        
+        // Here it errors with "This call is missing an argument of type (~a: int)"
+        cleanup 
+      })
+    }
+  ```
+* If you wanted explicitly uncurry a function, you needed to annotate it with the uncurried dot. 
+  ```rescript
+  (. param) => ()
+  ```
+* As ReScript could not fully statically analyze when to automatically uncurry a function over multiple files, it led to unnecessary `Curry.` calls in the emitted JavaScript code. 
```

Comment:
maybe we can clearly state here than function calls in uncurried mode are now guarantied to get compiled as simple JS function calls, which is quite nice for readability of the generated JS code but also for performance (even though it was not that significant) and sometimes even correctness when applied on bindings (`Curry` sometimes badly played with  `this` and other JS features we don't care about in rescript). 

Not sure it's worth being publicized but now you can even bind classes to simple rescript records which are sometimes much nicer to use than using an opaque type and a function:
```res
myLogger->Logger.info("foo")
// vs 
myLogger.info("foo")
```

Reply:
Thanks. I am not sure about the class suggestion either. I only added your first point. 
> function calls in uncurried mode are now guarantied to get compiled as simple JS function calls

Reply:
Can you maybe give me a full example?

Reply:
` (f) => f()`

is an example.
Or a more realistic example using that pattern.

Reply:
Sorry, I should have clarified I meant for classes as records.